### PR TITLE
6535 progress percent bug

### PR DESCRIPTION
**Pull request by linesarefuzzy**

### CODE REVIEW

File: app/models/accounting/transaction.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@
 #
 # Indexes
 #
-#  acc_trans_qbid_qbtype__unq_idx                          (qb_id,qb_transaction_type) UNIQUE
+#  acc_trans_qbid_qbtype_unq_idx                           (qb_id,qb_transaction_type) UNIQUE
```
Comment (hooverlunch) [inline app/models/accounting/transaction.rb]: Don't think this should be in here, can you please remove.

File: app/models/concerns/progress_calculable.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Methods for calculating progress towards completion of LoanResponseSets.
-# Assumes including classes implement required?, :group?, :answered?, and children.
+# Assumes including classes implement :required?, :group?, :answered?, :active?, and :children.
```
Comment (hooverlunch) [inline app/models/concerns/progress_calculable.rb]: Nice tidy.

File: app/models/concerns/progress_calculable.rb [hunk 3]
```diff
@@ -19,43 +19,38 @@ def progress_type
     required? ? "normal" : "optional"
   end
 
-  protected
+  # Inactive and retired questions should be ignored. Inactive questions only show when they are
+  # answered, and they are never required, so progress makes no sense. Retired questions should
+  # never show, so they should be excluded as well.
+  def active_children
+    children.select(&:active?)
```
Comment (hooverlunch) [inline app/models/concerns/progress_calculable.rb]: Good job using `select` and not firing a query. :)

File: app/models/concerns/progress_calculable.rb [hunk 4]
```diff
@@ -19,43 +19,38 @@ def progress_type
...
-  # plus the denominators of any child groups.
+  # plus the denominators of any required child groups.
   # If this is an optional node, the denominator is just the total number of child questions,
   # plus the denominators of any child groups.
   def progress_denominator
     return @progress_denominator if @progress_denominator
-    properties = {group: false, active: true}
-    properties[:required] = true if required?
+    return @progress_denominator = 0 unless active?
+    applicable_children = required? ? active_children.select(&:required?) : active_children
```
Comment (hooverlunch) [inline app/models/concerns/progress_calculable.rb]: This line is repeated. Move into private method?

File: app/models/concerns/progress_calculable.rb [hunk 5]
```diff
@@ -19,43 +19,38 @@ def progress_type
...
 
-    @progress_denominator = children.sum do |c|
-      (c.has_properties?(properties) ? 1 : 0) + c.progress_denominator
+    @progress_denominator = applicable_children.sum do |c|
+      (!c.group? ? 1 : 0) + c.progress_denominator
```
Comment (hooverlunch) [inline app/models/concerns/progress_calculable.rb]: Nice refactors overall.

File: db/schema.rb [hunk 6]
```diff
@@ -51,7 +51,7 @@
 
   add_index "accounting_transactions", ["accounting_account_id"], name: "index_accounting_transactions_on_accounting_account_id", using: :btree
   add_index "accounting_transactions", ["project_id"], name: "index_accounting_transactions_on_project_id", using: :btree
-  add_index "accounting_transactions", ["qb_id", "qb_transaction_type"], name: "acc_trans_qbid_qbtype__unq_idx", unique: true, using: :btree
+  add_index "accounting_transactions", ["qb_id", "qb_transaction_type"], name: "acc_trans_qbid_qbtype_unq_idx", unique: true, using: :btree
```
Comment (hooverlunch) [inline db/schema.rb]: Remove this also please.

File: spec/models/loan_response_progress_spec.rb [hunk 7]
```diff
@@ -4,109 +4,159 @@
...
+
+      resp = rset.response("f6")
+      expect(resp.progress_numerator).to eq 0
+      expect(resp.progress_denominator).to eq 0
+      expect(resp.progress).to eq 0
+    end
+
+    it "should be correct for the full custom value set" do
+      # Direct children contribute 0/1, group f3 contributes 2/3, total = 2/4 = 50%
+      # Optional, inactive and retired groups and questions should be ignored
```
Comment (hooverlunch) [inline spec/models/loan_response_progress_spec.rb]: Oh yeah

File: spec/models/loan_response_progress_spec.rb [hunk 8]
```diff
@@ -4,109 +4,159 @@
...
+      resp = rset.response("f4")
+      expect(resp.progress_numerator).to eq 1
+      expect(resp.progress_denominator).to eq 3
+      expect(resp.progress).to be_within(0.001).of(0.333)
+    end
+
+    it "should exclude inactive and retired questions from calculations" do
+      # Inactive questions only show when they are answered, and they are never required, so
+      # progress makes no sense. Retired questions should never show, so they should be excluded as
+      # well.
```
Comment (hooverlunch) [inline spec/models/loan_response_progress_spec.rb]: Drool

File: spec/models/loan_response_progress_spec.rb [hunk 9]
```diff
@@ -4,109 +4,159 @@
...
-  def create_question(parent: nil, name:, data_type:, required:)
+  def create_question(parent: nil, name:, data_type:, required:, status: 'active')
     field = create(:loan_question,
-      loan_question_set: set,
+      loan_question_set: qset,
       parent: parent,
       internal_name: name,
       data_type: data_type,
       override_associations: true,
+      status: status,
```
Comment (hooverlunch) [inline spec/models/loan_response_progress_spec.rb]: Stellar job on specs. Did you write them first? :P

**Comment by hooverlunch** [comment]
I'm not sure what that's about. Maybe @linesarefuzzy can comment. But I don't think it's a huge deal  for now...

**Comment by emitche** [comment]
Ok on the double underscore. It's been committed before with no problems so I doubt this will be an issue. Thanks.

**Pull request merged**